### PR TITLE
tests: Add proof-of-work fuzzing harness

### PR DESCRIPTION
**Pull request by practicalswift**
Add proof-of-work fuzzing harness.

### CODE REVIEW

**Comment by jonatack** [comment]
Concept ACK, commenting to be notified when this is rebased.

**Comment by practicalswift** [comment]
@jonasnick Rebased! :)

**Review by jonatack — COMMENTED** [review]
ACK baa778a

```
bitcoin/bitcoin (pr/18407)$ src/test/fuzz/pow
INFO: Seed: 3362987756
INFO: Loaded 1 modules   (514213 inline 8-bit counters): 514213 [0x562fb3600568, 0x562fb367de0d), 
INFO: Loaded 1 PC tables (514213 PCs): 514213 [0x562fb367de10,0x562fb3e56860), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2	INITED cov: 140 ft: 141 corp: 1/1b exec/s: 0 rss: 116Mb
#5	NEW    cov: 140 ft: 148 corp: 2/3b exec/s: 0 rss: 117Mb L: 2/2 MS: 3 ChangeByte-ShuffleBytes-CopyPart-
#8	NEW    cov: 140 ft: 153 corp: 3/6b exec/s: 0 rss: 117Mb L: 3/3 MS: 3 EraseBytes-InsertByte-InsertByte-
	NEW_FUNC[0/35]: 0x562fb06f9180 in boost::optional<CBlockHeader> ConsumeDeserializable<CBlockHeader>(FuzzedDataProvider&, unsigned long) /home/jon/projects/bitcoin/bitcoin/src/./test/fuzz/util.h:42
	NEW_FUNC[1/35]: 0x562fb06fcf30 in FuzzedDataProvider::Advance(unsigned long) /home/jon/projects/bitcoin/bitcoin/src/./test/fuzz/FuzzedDataProvider.h:247
#14	NEW    cov: 460 ft: 512 corp: 4/81b exec/s: 0 rss: 118Mb L: 75/75 MS: 1 InsertRepeatedBytes-
.../...
#220985	REDUCE cov: 1308 ft: 7563 corp: 309/508Kb exec/s: 373 rss: 517Mb L: 869/4096 MS: 3 ChangeASCIIInt-ChangeBit-EraseBytes-
#221115	REDUCE cov: 1308 ft: 7563 corp: 309/508Kb exec/s: 374 rss: 517Mb L: 147/4096 MS: 5 PersAutoDict-ChangeByte-CopyPart-EraseBytes-CopyPart- DE: "\x08\x00\x00\x00"-
```

**Comment by practicalswift** [comment]
@MarcoFalke Would you mind reviewing? :)

File: src/test/fuzz/pow.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#include <test/fuzz/fuzz.h>
+#include <test/fuzz/util.h>
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+void initialize()
+{
+    SelectParams(CBaseChainParams::REGTEST);
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/test/fuzz/pow.cpp:20]: or `         consensus.fPowNoRetargeting = true;`.

Those rule effectively bypass the code you want to fuzz. See for example here in line 51: https://doxygen.bitcoincore.org/pow_8cpp_source.html#l00049

**Review by maflcko — COMMENTED** [review]
ACK baa778ae1a7d75a448bafe30115f963e7de1fc18

File: src/test/fuzz/pow.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (practicalswift) [reply src/test/fuzz/pow.cpp]: Nice catch! Now using `CBaseChainParams::MAIN`.

**Comment by practicalswift** [comment]
@MarcoFalke Updated. Please re-review :)

**Pull request merged**